### PR TITLE
Add Vertx web extension

### PR DESCRIPTION
**Pull request by mkouba**
- make it possible to (1) define routes both in a declarative way and
programatically, (2) inject Router and observe Router during
http server startup

Supersedes https://github.com/quarkusio/quarkus/pull/1041

### CODE REVIEW

File: extensions/vertx-web/runtime/src/main/java/io/quarkus/vertx/web/runtime/HttpServerInitializer.java [hunk 1]
```diff
@@ -0,0 +1,160 @@
...
+            Thread.currentThread().interrupt();
+            throw new IllegalStateException("Unable to start the HTTP server", e);
+        }
+    }
+
+    // Note that we need a client proxy because if a bean also @Observes Router a null value would be injected 
+    @ApplicationScoped
+    @Produces
+    Router produceRouter() {
+        return router;
```
Comment (cescoffier) [inline extensions/vertx-web/runtime/src/main/java/io/quarkus/vertx/web/runtime/HttpServerInitializer.java]: We should produce the RX and Axle router also.

File: extensions/vertx-web/runtime/src/main/java/io/quarkus/vertx/web/runtime/HttpServerInitializer.java [hunk 2]
```diff
@@ -0,0 +1,160 @@
...
+    @Inject
+    Vertx vertx;
+
+    @Inject
+    Event<Router> routerEvent;
+
+    void initialize(VertxHttpConfiguration vertxHttpConfiguration, Map<String, List<Route>> routeHandlers,
+            LaunchMode launchMode) {
+        Router router = Router.router(vertx);
+        router.route().handler(BodyHandler.create());
```
Comment (cescoffier) [inline extensions/vertx-web/runtime/src/main/java/io/quarkus/vertx/web/runtime/HttpServerInitializer.java]: In dev mode, we should launch the scan.

File: bom/pom.xml [hunk 3]
```diff
@@ -1333,12 +1339,16 @@
...
-
             <dependency>
                 <groupId>io.vertx</groupId>
                 <artifactId>vertx-core</artifactId>
                 <version>${vertx.version}</version>
             </dependency>
+            <dependency>
+                <groupId>io.vertx</groupId>
+                <artifactId>vertx-web</artifactId>
+                <version>${vertx.version}</version>
```
Comment (cescoffier) [inline bom/pom.xml]: Need to depends also on the Axle and RX one.

**Review by cescoffier — APPROVED** [review]
I've made a few comments.

File: extensions/vertx-web/runtime/src/main/java/io/quarkus/vertx/web/runtime/HttpServerInitializer.java (hunk 1 above)
Comment (mkouba) [reply extensions/vertx-web/runtime/src/main/java/io/quarkus/vertx/web/runtime/HttpServerInitializer.java]: This is something I wanted to ask you @cescoffier. So I suppose that RX and Axle versions built on top of a regular router and we should follow the pattern from [VertxProducer](https://github.com/quarkusio/quarkus/blob/master/extensions/vertx/runtime/src/main/java/io/quarkus/vertx/runtime/VertxProducer.java#L43). But what's the advantage of the rx version... let's say for `Router`?

File: extensions/vertx-web/runtime/src/main/java/io/quarkus/vertx/web/runtime/HttpServerInitializer.java (hunk 2 above)
Comment (mkouba) [reply extensions/vertx-web/runtime/src/main/java/io/quarkus/vertx/web/runtime/HttpServerInitializer.java]: I don't understand ;-). What "scan"?

**Comment by mkouba** [comment]
@cescoffier I added the RX Router producer but I wasn't able to find the Axle version? Now it's also possible to change the `HttpServerOptions` in a CDI observer.

File: extensions/vertx-web/runtime/src/main/java/io/quarkus/vertx/web/runtime/HttpServerInitializer.java (hunk 2 above)
Comment (cescoffier) [reply extensions/vertx-web/runtime/src/main/java/io/quarkus/vertx/web/runtime/HttpServerInitializer.java]: I was referring to the hot reload. It should be possible to have a route triggering it.

File: extensions/vertx-web/runtime/src/main/java/io/quarkus/vertx/web/runtime/HttpServerInitializer.java (hunk 1 above)
Comment (cescoffier) [reply extensions/vertx-web/runtime/src/main/java/io/quarkus/vertx/web/runtime/HttpServerInitializer.java]: I agree that for the Router itself the RX API does not bring much. So, I won't mind not exposing it.

**Comment by cescoffier** [comment]
You are right we don't have the Axle API yet, my mistake.

File: extensions/vertx-web/runtime/src/main/java/io/quarkus/vertx/web/runtime/HttpServerInitializer.java (hunk 2 above)
Comment (mkouba) [reply extensions/vertx-web/runtime/src/main/java/io/quarkus/vertx/web/runtime/HttpServerInitializer.java]: I see. Yes, that would make sense.

File: extensions/vertx-web/runtime/src/main/java/io/quarkus/vertx/web/runtime/HttpServerInitializer.java (hunk 1 above)
Comment (mkouba) [reply extensions/vertx-web/runtime/src/main/java/io/quarkus/vertx/web/runtime/HttpServerInitializer.java]: Ok, we do produce the RX one now so let's keep it ;-)

**Comment by mkouba** [comment]
Hot replacement does not work yet. I'll need to ask Stuart about some magic he did for Undertow...

**Comment by cescoffier** [comment]
Merging like it is, we need to have a better approach for the hot-reload.

**Comment by mkouba** [comment]
@cescoffier well, it wasn't ready for merge yet ;-). But I think I know where the problem was and I'll try to send a follow-up PR this morning.

**Comment by murphye** [comment]
@cescoffier Very cool!

**Comment by cescoffier** [comment]
The hot reload has been added in  #1587

**Pull request merged**